### PR TITLE
ISPN-13672 Infinispan use cases

### PR DESCRIPTION
**Pull request by oraNod**
https://issues.redhat.com/browse/ISPN-13672

Backport to 13.0.x

### CODE REVIEW

File: documentation/src/main/asciidoc/topics/con_session_externalization.adoc [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+image::session-externalization-ispn.png[This illustration depicts a {brandname} cluster that offloads session data from applications.]
+endif::community[]
+//Downstream content
+ifdef::downstream[]
+image::session-externalization.png[This illustration depicts a {brandname} cluster that offloads session data from applications.]
+endif::downstream[]
+
+Externalizing sessions to {brandname} gives you the following benefits:
+
+Elasticity:: Make applications stateless to add and remove nodes without the need for rebalancing operations as well as to avoid downtime during cluster upgrades.
```
Comment (pferraro) [inline documentation/src/main/asciidoc/topics/con_session_externalization.adoc:19]: Clarification: externalizing sessions does not make the _application_ stateless, it makes the _application server_ stateless (or, at least, less stateful).  The core advantage is eliminating the need for rebalancing as a consequence of scaling the application server layer.  Generally, downtime should never an issue (provided the application server cluster is managed correctly)

File: documentation/src/main/asciidoc/topics/con_session_externalization.adoc [hunk 2]
```diff
@@ -0,0 +1,21 @@
+[id='session-externalization_{context}']
+= Session externalization
+
+{brandname} can provide external caches for applications built on Red Hat JBoss Enterprise Application Platform (EAP), Red Hat Build of Quarkus, Red Hat JBoss Web Server (JWS), and Spring.
```
Comment (dvagnero) [inline documentation/src/main/asciidoc/topics/con_session_externalization.adoc:4]: Since you name here the RH products I wonder if we should say Spring Boot instead or Spring Framework

File: documentation/src/main/asciidoc/topics/con_session_externalization.adoc [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+//Downstream content
+ifdef::downstream[]
+image::session-externalization.png[This illustration depicts a {brandname} cluster that offloads session data from applications.]
+endif::downstream[]
+
+Externalizing sessions to {brandname} gives you the following benefits:
+
+Elasticity:: Make applications stateless to add and remove nodes without the need for rebalancing operations as well as to avoid downtime during cluster upgrades.
+
+Smaller memory footprints:: Storing session data in external caches reduces memory pressure for applications, resulting in shorter and less frequent garbage collection (GC) runs.
```
Comment (pferraro) [inline documentation/src/main/asciidoc/topics/con_session_externalization.adoc:21]: On the contrary, externalizing sessions will generally result in more frequent GC runs, since session attribute objects will be created and discarded per request (rather than retained in memory across requests).  This can be mitigated by leveraging a near cache, in which case the heap requirements are comparable.

**Review by dvagnero — APPROVED** [review]
This looks amazing! Great work :)

**Review by pferraro — CHANGES_REQUESTED** [review]

File: documentation/src/main/asciidoc/topics/con_session_externalization.adoc (hunk 3 above)
Comment (oraNod) [reply documentation/src/main/asciidoc/topics/con_session_externalization.adoc:21]: Hmm... Thanks @pferraro I actually found this in some old JDG docs as I was trying to address why you'd want to externalize sessions in the first place. What if we drop "resulting in shorter and less frequent garbage collection (GC) runs."?

Does "storing session data in external caches reduces memory pressure for applications" make sense? Maybe "Storing session data in external caches reduces the overall amount of memory that applications require" is best?

File: documentation/src/main/asciidoc/topics/con_session_externalization.adoc (hunk 1 above)
Comment (oraNod) [reply documentation/src/main/asciidoc/topics/con_session_externalization.adoc:19]: Thanks again @pferraro I've reworded with this clarification.

File: documentation/src/main/asciidoc/topics/con_session_externalization.adoc (hunk 2 above)
Comment (oraNod) [reply documentation/src/main/asciidoc/topics/con_session_externalization.adoc]: Thanks for catching this. I've added variables for the community names. I think Spring is OK here. If we wanted to be sticklers I think we could maybe say Spring Framework but I'm not sure that's needed.

File: documentation/src/main/asciidoc/topics/con_in_memory_caching.adoc [hunk 4]
```diff
@@ -0,0 +1,21 @@
+[id='in-memory-caching_{context}']
+= In-memory caching
+
+{brandname} provides applications with data access that has much lower latency for read operations and higher throughput for write operations in comparison with traditional database systems.
+For in-memory caches, {brandname} stores key/value pairs in RAM only.
+
+.In-memory caches
+//Community content
+ifdef::community[]
+image::in-memory-cache-ispn.png[This illustration depicts a {brandname} cluster of four server nodes with in-memory caches that allow applications to perform read and write operations through remote clients.]
```
Comment (pruivo) [inline documentation/src/main/asciidoc/topics/con_in_memory_caching.adoc:10]: I'm failing to see what this image has to do with in-memory caches since it only shows the client-server architecture.

File: documentation/src/main/asciidoc/topics/con_in_memory_caching.adoc [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+endif::community[]
+//Downstream content
+ifdef::downstream[]
+image::in-memory-cache.png[This illustration depicts a {brandname} cluster of four server nodes with in-memory caches that allow applications to perform read and write operations through remote clients.]
+endif::downstream[]
+
+For in-memory caching deployments, {brandname}:
+
+* Maintains a single copy of each entry for higher performance read operations with lower durability guarantees.
+* Uses eviction or overflow to disk to avoid out of memory exceptions.
```
Comment (pruivo) [inline documentation/src/main/asciidoc/topics/con_in_memory_caching.adoc:20]: `Uses eviction or overflow`
Not sure if overflow is the right word here.

File: documentation/src/main/asciidoc/topics/con_in_memory_caching.adoc [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+image::in-memory-cache-ispn.png[This illustration depicts a {brandname} cluster of four server nodes with in-memory caches that allow applications to perform read and write operations through remote clients.]
+endif::community[]
+//Downstream content
+ifdef::downstream[]
+image::in-memory-cache.png[This illustration depicts a {brandname} cluster of four server nodes with in-memory caches that allow applications to perform read and write operations through remote clients.]
+endif::downstream[]
+
+For in-memory caching deployments, {brandname}:
+
+* Maintains a single copy of each entry for higher performance read operations with lower durability guarantees.
```
Comment (pruivo) [inline documentation/src/main/asciidoc/topics/con_in_memory_caching.adoc:19]: Is this related to the caching service in openshift?

File: documentation/src/main/asciidoc/topics/con_in_memory_caching.adoc [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+* Uses eviction or overflow to disk to avoid out of memory exceptions.
+* Clients use near-caching to avoid repeatable reads and gain performance.
```
Comment (pruivo) [inline documentation/src/main/asciidoc/topics/con_in_memory_caching.adoc:21]: `repeatable reads` <- can we avoid this combination of words? It may confuse users since it is a transactional consistency level :)

File: documentation/src/main/asciidoc/topics/con_side_caching.adoc [hunk 8]
```diff
@@ -0,0 +1,22 @@
+[id='side-caching_{context}']
+= Side caching
+
+{brandname} stores data that applications retrieve from persistent storage, which reduces the number of read operations to persistent storage and increases response times for subsequent reads.
+
+.Side caches
+//Community content
+ifdef::community[]
+image::side-cache-ispn.png[This illustration depicts a {brandname} side cache to which applications write to avoid repeatedly reading the same entries from persistent storage.]
```
Comment (pruivo) [inline documentation/src/main/asciidoc/topics/con_side_caching.adoc:9]: Wondering what is your opinion about:
* stage 1 - > add an arrow from the cache to  the application saying "not found"
* stage 2 -> split stage 2 into two stages: the first one reads from persistence, the second one stores in cache

**Review by pruivo — APPROVED** [review]
So, we are duplicating images just because of the words "Infinispan" and "Data Grid". Is it possible to do anything about it? Can we come up with a more generic word?

File: documentation/src/main/asciidoc/topics/con_in_memory_caching.adoc (hunk 4 above)
Comment (oraNod) [reply documentation/src/main/asciidoc/topics/con_in_memory_caching.adoc:10]: In a separate topic we mention that client/server is preferable to embedded because the server gives you logging and monitoring and separates the cache from application logic. This diagram is trying to be consistent with that message by showing remote caches in a dedicated JVM. So it's still in memory but not in the same memory space as code execution. I'll make some changes to the wording here but do you think this should be "In-memory caches in a client/server architecture" or should those two things be completely distinct?

File: documentation/src/main/asciidoc/topics/con_in_memory_caching.adoc (hunk 5 above)
Comment (oraNod) [reply documentation/src/main/asciidoc/topics/con_in_memory_caching.adoc:20]: Well spotted. Fixed.

File: documentation/src/main/asciidoc/topics/con_side_caching.adoc (hunk 8 above)
Comment (oraNod) [reply documentation/src/main/asciidoc/topics/con_side_caching.adoc:9]: Yeah, we could do that. I'll need to work with the illustrator so would like to do that in a separate PR. I've created https://issues.redhat.com/browse/ISPN-13701 so we don't lose track of it.

File: documentation/src/main/asciidoc/topics/con_in_memory_caching.adoc (hunk 7 above)
Comment (oraNod) [reply documentation/src/main/asciidoc/topics/con_in_memory_caching.adoc:21]: Fixed

File: documentation/src/main/asciidoc/topics/con_in_memory_caching.adoc (hunk 6 above)
Comment (oraNod) [reply documentation/src/main/asciidoc/topics/con_in_memory_caching.adoc:19]: Yes, I have to confess I'm not 100% sure how this slipped in here. I've removed it. Thanks.

**Comment by oraNod** [comment]
> So, we are duplicating images just because of the words "Infinispan" and "Data Grid". Is it possible to do anything about it? Can we come up with a more generic word?

I'm going through and replacing the *-ispn.png images with SVG format that we can use in community. We're restricted to PNG downstream however. I get that it's not ideal to duplicate images for the names but I don't know about a generic alternative here. 

SVG is just XML under the hood so you can s/Infinispan/Data Grid but I'm not sure if it's possible to do that and then generate PNG automatically during the sync. It seems like one way but a lot of trouble to do that every time.

File: documentation/src/main/asciidoc/topics/con_in_memory_caching.adoc (hunk 4 above)
Comment (oraNod) [reply documentation/src/main/asciidoc/topics/con_in_memory_caching.adoc]: @pruivo Excellent comment, as always. I've actually just dropped the "in-memory" topic. We can include the diagram as part of the deployment model maybe. Or we could drop it altogether if it's not adding value.

**Review by karesti — APPROVED** [review]

**Review by karesti — APPROVED** [review]
